libraries/AP_HAL_Linux/CAN.cpp

```diff
@@ -0,0 +1,619 @@
+/*
+ * Copyright (C) 2014-2015 Pavel Kirienko <pavel.kirienko@gmail.com>
+ *                         Ilia Sheremet <illia.sheremet@gmail.com>
+ *
+ * Modification for Ardupilot
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_HAL/system.h>
+
+#include <AP_BoardConfig/AP_BoardConfig.h>
+
+#if HAL_WITH_UAVCAN
+
+#include "CAN.h"
+
+#include <AP_UAVCAN/AP_UAVCAN.h>
+
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <sys/socket.h>
+#include <net/if.h>
+#include <linux/can/raw.h>
+
+extern const AP_HAL::HAL& hal;
+
+using namespace Linux;
+
+uavcan::MonotonicTime getMonotonic()
+{
+    return uavcan::MonotonicTime::fromUSec(AP_HAL::micros64());
+}
+
+static can_frame makeSocketCanFrame(const uavcan::CanFrame& uavcan_frame)
+{
+    can_frame sockcan_frame { uavcan_frame.id& uavcan::CanFrame::MaskExtID, uavcan_frame.dlc, { } };
+    std::copy(uavcan_frame.data, uavcan_frame.data + uavcan_frame.dlc, sockcan_frame.data);
+    if (uavcan_frame.isExtended()) {
+        sockcan_frame.can_id |= CAN_EFF_FLAG;
+    }
+    if (uavcan_frame.isErrorFrame()) {
+        sockcan_frame.can_id |= CAN_ERR_FLAG;
+    }
+    if (uavcan_frame.isRemoteTransmissionRequest()) {
+        sockcan_frame.can_id |= CAN_RTR_FLAG;
+    }
+    return sockcan_frame;
+}
+
+static uavcan::CanFrame makeUavcanFrame(const can_frame& sockcan_frame)
+{
+    uavcan::CanFrame uavcan_frame(sockcan_frame.can_id & CAN_EFF_MASK, sockcan_frame.data, sockcan_frame.can_dlc);
+    if (sockcan_frame.can_id & CAN_EFF_FLAG) {
+        uavcan_frame.id |= uavcan::CanFrame::FlagEFF;
+    }
+    if (sockcan_frame.can_id & CAN_ERR_FLAG) {
+        uavcan_frame.id |= uavcan::CanFrame::FlagERR;
+    }
+    if (sockcan_frame.can_id & CAN_RTR_FLAG) {
+        uavcan_frame.id |= uavcan::CanFrame::FlagRTR;
+    }
+    return uavcan_frame;
+}
+
+bool CAN::begin(uint32_t bitrate)
+{
+    if (_initialized) return _initialized;
+    // TODO: Add possibility change bitrate
+    _fd = openSocket(HAL_BOARD_CAN_IFACE_NAME);
+    if (_fd > 0) {
+        _bitrate = bitrate;
+        _initialized = true;
+    } else {
+        _initialized = false;
+    }
+    return _initialized;
+}
+
+void CAN::reset()
+{
+    if (_initialized && _bitrate != 0) {
+        close(_fd);
+        begin(_bitrate);
+    }
+}
+
+void CAN::end()
+{
+    _initialized = false;
+    close(_fd);
+}
+
+bool CAN::is_initialized()
+{
+    return _initialized;
+}
+
+int32_t CAN::tx_pending()
+{
+    if (_initialized) {
+        return _tx_queue.size();
+    } else {
+        return -1;
+    }
+}
+
+int32_t CAN::available()
+{
+    if (_initialized) {
+        return _rx_queue.size();
+    } else {
+        return -1;
+    }
+}
+
+int CAN::openSocket(const std::string& iface_name)
+{
+    errno = 0;
+
+    int s = socket(PF_CAN, SOCK_RAW, CAN_RAW);
+    if (s < 0) {
+        return s;
+    }
+
+    std::shared_ptr<int> fd_guard(new int(s), [](const int *fd) {
```

Comment:
@lucasdemarchi I don't see your comment anymore, did you remove it? Is this line ok as-is or are you still suggesting it change to be like 
https://github.com/intel/mavlink-router/blob/master/src/mavlink-router/logendpoint.cpp#L105
which is 
`std::shared_ptr<void> defer(dir, [](DIR *p) { closedir(p); });`

Reply:
It's still there, on the previous PR: https://github.com/ArduPilot/ardupilot/pull/6398#discussion_r138697422

Reply:
ok, I see it. Anything else needed on this PR? I'd love to get this merged (when ready)

Reply:
I'll happily update the elif stuff myself, just wondering about the linux/CAN specific stuff